File: gui/pitcrewgui/ansi.py
"""ANSI escape sequences in a log file, turned into spans a TextView can style.

A dev server's log is not plain text. Spring Boot, Vite, gradle, npm and pip all
write SGR colour into it, and pitcrew captures stdout verbatim — as it should,
because the file is meant to be readable with `less -R` too. The desktop app
was rendering those bytes literally, so a Spring log came out as a wall of
`▯▯[2m2026-08-20 11:04:19.670▯▯[0;39m ▯▯[32mINFO▯▯[0;39m`, with the actual
message pushed off the right-hand edge.

The terminal viewer solves this by throwing the colour away (`strip_ansi`).
That is the right call for a frame it has to redraw at a fixed width, and the
wrong one here: those colours are how you find the WARN in three hundred lines
of INFO, and a window has a tag table to spend on them.

No GTK in this file, so the parsing rules are testable without a display.
"""

from __future__ import annotations

import re

# Every escape sequence, not only the ones we act on. CSI (\\x1b[...), OSC
# (\\x1b]...BEL or ST), and the two-character sequences. Anything matched and
# not understood is dropped rather than printed — an erase-line or a cursor
# move has no meaning in a scrollback buffer, but it is still not text.
_SEQ = re.compile(
    r"\x1b\[([0-9;:]*)m"                 # 1: SGR, the only kind we interpret
    r"|\x1b\[[0-9;:?]*[ -/]*[@-~]"       # other CSI
    r"|\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)"  # OSC (window titles)
    r"|\x1b[ -/]+[0-~]"                  # nF, e.g. ESC ( B — charset selection
    r"|\x1b[@-Z\\-_]"                    # two-character escapes
)
_CONTROL = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")

# SGR foreground codes to a name the view maps onto a palette. Bright variants
# are separate because "bright black" is a real grey that logs use for
# timestamps, and folding it into black would make those invisible.
_FG = {
    30: "black", 31: "red", 32: "green", 33: "yellow",
    34: "blue", 35: "magenta", 36: "cyan", 37: "white",
    90: "bright-black", 91: "bright-red", 92: "bright-green", 93: "bright-yellow",
    94: "bright-blue", 95: "bright-magenta", 96: "bright-cyan", 97: "bright-white",
}
_ATTR = {1: "bold", 2: "dim", 3: "italic", 4: "underline"}
_ATTR_OFF = {22: ("bold", "dim"), 23: ("italic",), 24: ("underline",)}

# xterm's 6x6x6 cube and greyscale ramp, for 38;5;N. Logs that use 256-colour
# mostly use it for greys and pastels, and rounding those to the nearest of
# eight would lose the distinction entirely.
_CUBE = (0, 95, 135, 175, 215, 255)


def _xterm256(index: int) -> str | None:
    if index < 16:
        return None                       # a palette colour; the caller names it
    if index < 232:
        index -= 16
        r, g, b = _CUBE[index // 36], _CUBE[index % 36 // 6], _CUBE[index % 6]
        return f"#{r:02x}{g:02x}{b:02x}"
    grey = 8 + (index - 232) * 10
    return f"#{grey:02x}{grey:02x}{grey:02x}"


class Style:
    """Foreground and attributes at a point in the stream."""

    __slots__ = ("fg", "attrs")

    def __init__(self, fg: str | None = None, attrs: frozenset[str] = frozenset()):
        self.fg = fg
        self.attrs = attrs

    def tags(self) -> tuple[str, ...]:
        return ((f"fg:{self.fg}",) if self.fg else ()) + tuple(sorted(self.attrs))

    def __eq__(self, other) -> bool:
        return isinstance(other, Style) and self.fg == other.fg and self.attrs == other.attrs

    def __hash__(self) -> int:
        return hash((self.fg, self.attrs))


def _apply(style: Style, params: str) -> Style:
    """One SGR sequence against the current style."""
    # "\\x1b[m" means "\\x1b[0m". An empty parameter inside a list means 0 too.
    codes = [int(p) if p.isdigit() else 0
             for p in (params.split(";") if params else ["0"])]
    fg, attrs = style.fg, set(style.attrs)
    index = 0
    while index < len(codes):
        code = codes[index]
        if code == 0:
            fg, attrs = None, set()
        elif code in _ATTR:
            attrs.add(_ATTR[code])
        elif code in _ATTR_OFF:
            attrs.difference_update(_ATTR_OFF[code])
        elif code in _FG:
            fg = _FG[code]
        elif code == 39:
            fg = None
        elif code == 38 and index + 1 < len(codes):
            # 38;5;N and 38;2;R;G;B. Consumed here so their arguments are never
            # mistaken for further codes — that is how a truecolor sequence
            # ends up setting a random background in a naive parser.
            kind = codes[index + 1]
            if kind == 5 and index + 2 < len(codes):
                named = _FG.get(codes[index + 2] + (30 if codes[index + 2] < 8 else 82))
                fg = named or _xterm256(codes[index + 2])
                index += 2
            elif kind == 2 and index + 4 < len(codes):
                r, g, b = codes[index + 2:index + 5]
                fg = f"#{r & 255:02x}{g & 255:02x}{b & 255:02x}"
                index += 4
        # Backgrounds (40-49, 100-107) are recognised by the code above only to
        # be skipped. A log that paints whole lines is unreadable in a window
        # that is not a terminal, and losing that emphasis beats a wall of
        # colour blocks.
        index += 1
    return Style(fg, frozenset(attrs))


def spans(line: str) -> list[tuple[str, tuple[str, ...]]]:
    """(text, tag names) pairs for one line, escapes consumed.

    Style does NOT carry across lines. It can in a real terminal, but a log
    buffer is scrolled, filtered and trimmed — one unterminated sequence would
    otherwise colour everything after it, including lines the writer had
    already reset. Per-line is the reading that survives being cut up.
    """
    line = _collapse_carriage_returns(line)
    out: list[tuple[str, tuple[str, ...]]] = []
    style = Style()
    position = 0
    for match in _SEQ.finditer(line):
        text = line[position:match.start()]
        if text:
            out.append((_CONTROL.sub("", text), style.tags()))
        if match.group(1) is not None:
            style = _apply(style, match.group(1))
        position = match.end()
    text = line[position:]
    if text:
        out.append((_CONTROL.sub("", text), style.tags()))
    return [(text, tags) for text, tags in out if text]


def _collapse_carriage_returns(line: str) -> str:
    """What a terminal would leave on screen after the line was overwritten.

    npm, pip and gradle draw progress by returning to column zero and writing
    again. Kept verbatim, one download turns into two hundred copies of itself.
    """
    if "\r" not in line:
        return line
    return line.rsplit("\r", 1)[-1]


def plain(line: str) -> str:
    """The line as text: what a filter should search and a pattern should match."""
    return "".join(text for text, _tags in spans(line))

File: crates/pitcrew-core/src/ansi.rs
//! ANSI escapes in a log, turned into styled spans.
//!
//! **A dev server's log is not plain text.** Spring Boot, Vite, gradle, npm and
//! pip all write SGR colour into it, and pitcrew captures stdout verbatim — as
//! it should, because the file is meant to be readable with `less -R` too.
//! Rendering those bytes literally turns a Spring log into a wall of
//! `␛[2m2026-08-20 11:04:19␛[0;39m ␛[32mINFO␛[0;39m` with the actual message
//! pushed off the right-hand edge.
//!
//! The terminal viewer solves this by throwing the colour away. That is the
//! right call for a frame it redraws at a fixed width and the wrong one for a
//! window: those colours are how you find the WARN in three hundred lines of
//! INFO.
//!
//! **This is a parser, not a stripper.** Do not "simplify" it into one. And no
//! GTK here, so the rules are testable without a display.

/// One run of text that shares a style.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Span {
    pub text: String,
    pub style: Style,
}

/// What SGR asked for. Deliberately a small subset: the attributes a log
/// actually uses, and nothing else.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct Style {
    /// The 16-colour index, or `None` for the default foreground.
    pub fg: Option<u8>,
    pub bold: bool,
    /// SGR 2. Spring Boot writes its timestamps this way, and rendering them
    /// at full strength is most of why its logs look like noise.
    pub dim: bool,
    pub italic: bool,
    pub underline: bool,
}

/// Split a line into styled spans.
///
/// Every escape sequence is consumed, not only the ones acted on: an
/// erase-line or a cursor move has no meaning in a scrollback buffer, but it is
/// still not text and printing it is worse than dropping it.
pub fn parse(line: &str) -> Vec<Span> {
    let mut out: Vec<Span> = Vec::new();
    let mut style = Style::default();
    let mut text = String::new();
    let mut chars = line.chars().peekable();

    while let Some(c) = chars.next() {
        if c != '\u{1b}' {
            text.push(c);
            continue;
        }
        // Flush what came before the escape under the style it had.
        if !text.is_empty() {
            out.push(Span {
                text: std::mem::take(&mut text),
                style,
            });
        }
        match chars.peek() {
            // CSI — the only kind carrying anything worth keeping.
            Some('[') => {
                chars.next();
                let mut params = String::new();
                let mut final_byte = None;
                for c in chars.by_ref() {
                    if c.is_ascii_alphabetic() {
                        final_byte = Some(c);
                        break;
                    }
                    params.push(c);
                }
                if final_byte == Some('m') {
                    apply(&mut style, &params);
                }
                // Anything else (cursor moves, erases) is consumed and dropped.
            }
            // OSC — a window title, usually. Runs to BEL or ST.
            Some(']') => {
                chars.next();
                let mut prev = '\0';
                for c in chars.by_ref() {
                    if c == '\u{7}' || (prev == '\u{1b}' && c == '\\') {
                        break;
                    }
                    prev = c;
                }
            }
            // Everything else. ESC followed by an INTERMEDIATE byte (0x20-0x2F)
            // takes a final byte after it — `ESC ( B` is three bytes, not two,
            // and consuming only two leaves the `B` in the text as a stray
            // letter in the middle of a log line.
            Some(_) => {
                while chars
                    .peek()
                    .is_some_and(|c| ('\u{20}'..='\u{2f}').contains(c))
                {
                    chars.next();
                }
                // The final byte.
                chars.next();
            }
            None => {}
        }
    }
    if !text.is_empty() {
        out.push(Span { text, style });
    }
    out
}

fn apply(style: &mut Style, params: &str) {
    // A bare `\x1b[m` means reset, same as `\x1b[0m`.
    if params.is_empty() {
        *style = Style::default();
        return;
    }
    let mut codes = params.split(';').peekable();
    while let Some(raw) = codes.next() {
        // An empty field is a zero: `\x1b[0;39m` and `\x1b[;39m` mean the same.
        let code: u16 = raw.trim().parse().unwrap_or(0);
        match code {
            0 => *style = Style::default(),
            1 => style.bold = true,
            2 => style.dim = true,
            3 => style.italic = true,
            4 => style.underline = true,
            21 | 22 => {
                style.bold = false;
                style.dim = false;
            }
            23 => style.italic = false,
            24 => style.underline = false,
            30..=37 => style.fg = Some((code - 30) as u8),
            90..=97 => style.fg = Some((code - 90 + 8) as u8),
            39 => style.fg = None,
            // 256-colour and truecolour. Only the 256-colour form maps onto the
            // palette here; truecolour is consumed so its arguments are not
            // read back as separate attributes.
            38 => match codes.next().and_then(|m| m.parse::<u16>().ok()) {
                Some(5) => {
                    style.fg = codes.next().and_then(|v| v.parse::<u8>().ok());
                }
                Some(2) => {
                    codes.next();
                    codes.next();
                    codes.next();
                }
                _ => {}
            },
            _ => {}
        }
    }
}

/// The same line with every escape removed — for a fixed-width frame that has
/// no tag table to spend on colour.
pub fn strip(line: &str) -> String {
    parse(line).into_iter().map(|s| s.text).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn texts(line: &str) -> Vec<String> {
        parse(line).into_iter().map(|s| s.text).collect()
    }

    #[test]
    fn plain_text_is_one_span() {
        assert_eq!(
            parse("hello"),
            [Span {
                text: "hello".into(),
                style: Style::default()
            }]
        );
        assert!(parse("").is_empty());
    }

    /// The shape that made a Spring log unreadable.
    #[test]
    fn a_spring_boot_line_comes_apart_into_readable_spans() {
        let line = "\u{1b}[2m2026-08-20 11:04:19\u{1b}[0;39m \u{1b}[32mINFO\u{1b}[0;39m started";
        let spans = parse(line);
        assert_eq!(spans[0].text, "2026-08-20 11:04:19");
        assert!(
            spans[0].style.dim,
            "the timestamp is dim, which is why it reads as noise at full strength"
        );
        let info = spans.iter().find(|s| s.text == "INFO").expect("the level");
        assert_eq!(info.style.fg, Some(2), "green");
        assert!(!info.style.dim, "the reset cleared it");
        assert_eq!(strip(line), "2026-08-20 11:04:19 INFO started");
    }

    /// `\x1b[0;39m` and `\x1b[;39m` mean the same thing, and a parser that
    /// reads an empty field as anything but zero drops the reset.
    #[test]
    fn an_empty_parameter_is_a_zero() {
        let a = parse("\u{1b}[1mx\u{1b}[0;39my");
        let b = parse("\u{1b}[1mx\u{1b}[;39my");
        assert_eq!(a, b);
        assert!(!a[1].style.bold);
    }

    /// A bare `\x1b[m` is a reset.
    #[test]
    fn a_bare_sgr_is_a_reset() {
        let spans = parse("\u{1b}[1;31mloud\u{1b}[mquiet");
        assert!(spans[0].style.bold && spans[0].style.fg == Some(1));
        assert_eq!(spans[1].style, Style::default());
    }

    /// An erase-line or a cursor move has no meaning in a scrollback buffer,
    /// but it is still not text.
    #[test]
    fn sequences_that_are_not_colour_are_dropped_rather_than_printed() {
        assert_eq!(strip("a\u{1b}[2Kb"), "ab");
        assert_eq!(strip("a\u{1b}[1;1Hb"), "ab");
        // OSC: a window title, to BEL or to ST.
        assert_eq!(strip("a\u{1b}]0;title\u{7}b"), "ab");
        assert_eq!(strip("a\u{1b}]0;title\u{1b}\\b"), "ab");
        // A two-character sequence.
        assert_eq!(strip("a\u{1b}(Bb"), "ab");
    }

    #[test]
    fn bright_colours_land_in_the_upper_half_of_the_palette() {
        assert_eq!(parse("\u{1b}[91mx")[0].style.fg, Some(9));
        assert_eq!(parse("\u{1b}[31mx")[0].style.fg, Some(1));
    }

    /// Truecolour is consumed whole. Reading its three arguments back as
    /// separate attributes is how a red foreground became bold-and-underlined.
    #[test]
    fn truecolour_arguments_are_not_read_as_attributes() {
        let spans = parse("\u{1b}[38;2;1;2;4mx");
        assert!(!spans[0].style.bold, "the 1 was a colour component");
        assert!(!spans[0].style.underline, "so was the 4");
    }

    #[test]
    fn a_256_colour_index_is_kept() {
        assert_eq!(parse("\u{1b}[38;5;208mx")[0].style.fg, Some(208));
    }

    /// A log line can end mid-sequence, because it is being appended to while
    /// it is read.
    #[test]
    fn an_unterminated_escape_does_not_leak_into_the_text() {
        assert_eq!(strip("done\u{1b}[3"), "done");
        assert_eq!(strip("done\u{1b}"), "done");
    }

    #[test]
    fn text_is_preserved_exactly_between_escapes() {
        assert_eq!(texts("\u{1b}[31ma b  c\u{1b}[0m"), ["a b  c"]);
        // Including characters that look like escapes but are not.
        assert_eq!(strip("cost: [30%]"), "cost: [30%]");
    }
}
